modules/calib3d/src/ptsetreg.cpp

```diff
@@ -900,6 +900,86 @@ int estimateAffine3D(InputArray _from, InputArray _to,
     return createRANSACPointSetRegistrator(makePtr<Affine3DEstimatorCallback>(), 4, ransacThreshold, confidence)->run(dFrom, dTo, _out, _inliers);
 }
 
+Mat    estimateAffine3D(InputArray _from, InputArray _to,
+                        CV_OUT double* _scale, bool force_rotation)
+{
+    CV_INSTRUMENT_REGION();
+    Mat from = _from.getMat(), to = _to.getMat();
+    int count = from.checkVector(3);
+
+    CV_CheckGE(count, 4, "Umeyama algorithm needs at least 4 points for affine transformation estimation.");
+    CV_CheckEQ(to.checkVector(3), count, "Point sets need to have the same size");
+    from = from.reshape(1, count);
+    to = to.reshape(1, count);
+    if(from.type() != CV_64F)
+        from.convertTo(from, CV_64F);
+    if(to.type() != CV_64F)
+        to.convertTo(to, CV_64F);
+
+    const double one_over_n = 1./count;
+
+    const auto colwise_mean = [one_over_n](const Mat& m)
+    {
+        Mat my;
+        reduce(m, my, 0, REDUCE_SUM, CV_64F);
+        return my * one_over_n;
+    };
+
+    const auto demean = [count](const Mat& A, const Mat& mean)
+    {
+        Mat A_centered = Mat::zeros(count, 3, CV_64F);
+        for(int i = 0; i < count; i++)
+        {
+            A_centered.row(i) = A.row(i) - mean;
+        }
+        return A_centered;
+    };
+
+    Mat from_mean = colwise_mean(from);
+    Mat to_mean = colwise_mean(to);
+
+    Mat from_centered = demean(from, from_mean);
+    Mat to_centered = demean(to, to_mean);
+
+    Mat cov = to_centered.t() * from_centered * one_over_n;
+
+    Mat u,d,vt;
+    SVD::compute(cov, d, u, vt, SVD::MODIFY_A | SVD::FULL_UV);
+
+    CV_CheckGE(countNonZero(d), 2, "Points cannot be colinear");
```

Comment:
this check may throw exception, which is not a good idea for a numerical algorithm. Instead, it should provide some reasonable response, i.e. return empty matrix or maybe one of possible solutions (if there are many).

Reply:
If the points are collinear the problem is underconstrained and there is an infinite number of solutions. However it is common to throw on malformed arguments, for example if the input matrices are not the correct dimension.
E.g. `CV_CheckEQ(to.checkVector(3), count, "Point sets need to have the same size");` I would consider this to be a similar case.

I would suggest extending the documentation to saying that co-linear points are not valid. Returning a possible solution would involve matching two two point sets, of which at least one lies on a line. So we would need to choose a random plane to project to to do so. It would involve quite some heuristics and would be pretty hard for the user to unravel if it goes wrong.

BTW.: I tried what would happen with the existing implementation:

``` 
import cv2
import numpy as np

a = np.array([[1., 1, 1, 1],[0, 0, 0, 0],[0, 0, 0, 0]]).T

r, o, i = cv2.estimateAffine3D(a, a)
print(o, r)
```

returns:

```
[[0.5 0.  0.  0.5]
 [0.  0.  0.  0. ]
 [0.  0.  0.  0. ]] 1
```

Should I maybe open an issue, at least this should probably not report a successful execution.